### PR TITLE
Support for cracking InnoSetup generated Unicode installers

### LINKED ISSUE #5318: Derivate of dynamic_25 with UTF-16 Unicode support
In order to recover passwords built with the Unicode builds of the well-known installer InnoSetup, it is required that a hash is found that is constructed using `sha1(salt_binary || password_utf16)`.

For ANSI builds of InnoSetup, the scheme `sha1(salt_binary || password_ansi)` (which is `$dynamic_25$`) works fine.

My suggestion is that a new `$dynamic_xx$` is added to the main distribution, since I think it might be useful for many.
Thank you very much.

### PR DESCRIPTION
**Pull request by kholia**
Closes: https://github.com/openwall/john/issues/5318.

Related: https://github.com/kholia/inno2john/tree/tweaks-v3.

### CODE REVIEW

File: src/dynamic_preloads.c [hunk 1]
```diff
@@ -3302,6 +3326,7 @@ static DYNAMIC_Setup Setups[] =
 	{ "dynamic_23: sha1(md5($p))",              _Funcs_23,_Preloads_23,_ConstDefault, MGF_NO_FLAG, MGF_INPUT_20_BYTE|MGF_KEYS_INPUT },
 	{ "dynamic_24: sha1($p.$s)",                _Funcs_24,_Preloads_24,_ConstDefault, MGF_FLAT_BUFFERS|MGF_SALTED, MGF_NO_FLAG|MGF_INPUT_20_BYTE, -64, 110, 110 },
 	{ "dynamic_25: sha1($s.$p)",                _Funcs_25,_Preloads_25,_ConstDefault, MGF_FLAT_BUFFERS|MGF_SALTED, MGF_NO_FLAG|MGF_INPUT_20_BYTE, -64, 110, 110 },
+	{ "dynamic_41: sha1($s.utf16($p))",         _Funcs_41,_Preloads_41,_ConstDefault, MGF_UTF8|MGF_FLAT_BUFFERS|MGF_SALTED, MGF_NO_FLAG|MGF_INPUT_20_BYTE, -64, 110, 110 },
```
Comment (kholia) [inline src/dynamic_preloads.c]: I am not sure about `-64, 110, 110` stuff here.

**Review by ghost — APPROVED** [review]
We should merge this so people can test it, but I bet 110 is wrong.

Since it is UFT-16 (I mean, you are setting the UTF-8 flag), 110 seems unsafe.
I think ..., 27, 40 is the first safe choice we should use (meanwhile). Anyway,  If it passes self-test you can merge as is.

**Comment by solardiz** [comment]
Thank you @kholia! I'm sorry I didn't comment sooner. 110 definitely looks wrong, and also `MGF_UTF8` is questionable - maybe @magnumripper can comment on these before we merge?

**Comment by magnumripper** [comment]
> Thank you @kholia! I'm sorry I didn't comment sooner. 110 definitely looks wrong, and also `MGF_UTF8` is questionable - maybe @magnumripper can comment on these before we merge?

MGF_UTF8 is likely correct, and I'm not 100% sure but if it can do 110 for ASCII it can do 55 for UTF-16. So maybe 55, 55?

**Comment by kholia** [comment]
Thanks for the help!

I have pushed the following change:

```diff
diff --git a/src/dynamic_preloads.c b/src/dynamic_preloads.c
index 84f7d10c3..fc3c415b1 100644
--- a/src/dynamic_preloads.c
+++ b/src/dynamic_preloads.c
@@ -3326,7 +3326,7 @@ static DYNAMIC_Setup Setups[] =
        { "dynamic_23: sha1(md5($p))",              _Funcs_23,_Preloads_23,_ConstDefault, MGF_NO_FLAG, MGF_INPUT_20_BYTE|MGF_KEYS_INPUT },
        { "dynamic_24: sha1($p.$s)",                _Funcs_24,_Preloads_24,_ConstDefault, MGF_FLAT_BUFFERS|MGF_SALTED, MGF_NO_FLAG|MGF_INPUT_20_BYTE, -64, 110, 110 },
        { "dynamic_25: sha1($s.$p)",                _Funcs_25,_Preloads_25,_ConstDefault, MGF_FLAT_BUFFERS|MGF_SALTED, MGF_NO_FLAG|MGF_INPUT_20_BYTE, -64, 110, 110 },
-       { "dynamic_41: sha1($s.utf16($p))",         _Funcs_41,_Preloads_41,_ConstDefault, MGF_UTF8|MGF_FLAT_BUFFERS|MGF_SALTED, MGF_NO_FLAG|MGF_INPUT_20_BYTE, -64, 110, 110 },
+       { "dynamic_41: sha1($s.utf16($p))",         _Funcs_41,_Preloads_41,_ConstDefault, MGF_UTF8|MGF_FLAT_BUFFERS|MGF_SALTED, MGF_NO_FLAG|MGF_INPUT_20_BYTE, -64, 55, 55 },
```

**Pull request merged**